crates/paymaster-execution/src/diagnostics/service.rs

```diff
@@ -0,0 +1,389 @@
+//! Diagnostic service that manages extractors and orchestrates error analysis.
+
+use super::context::DiagnosticContext;
+use super::extractor::{CallDiagnostic, CallMetadataExtractor};
+use crate::tokens::TokenService;
+use starknet::core::types::Felt;
+use std::sync::Arc;
+use tracing::{debug, info_span, warn};
+
+/// Service that manages a registry of metadata extractors and analyzes transaction errors.
+///
+/// The service iterates through registered extractors to find one that can handle
+/// the given error context, then extracts and logs diagnostic information.
+#[derive(Clone)]
+pub struct DiagnosticService {
+    extractors: Vec<Arc<dyn CallMetadataExtractor>>,
+}
+
+impl DiagnosticService {
+    /// Creates a diagnostic service pre-configured with all known extractors.
+    ///
+    /// Currently includes:
+    /// - AVNU Exchange extractor (mainnet address)
+    pub fn new(chain_id: Felt) -> Self {
+        let token_service = TokenService::for_chain(chain_id);
+        use super::extractors::{AvnuExtractor, AVNU_EXCHANGE_ADDRESS};
+
+        Self::empty().with_extractor(AvnuExtractor::with_token_service(AVNU_EXCHANGE_ADDRESS, token_service))
+    }
+
+    /// Creates an empty diagnostic service with no extractors.
+    #[cfg(test)]
+    pub(crate) fn empty() -> Self {
+        Self { extractors: Vec::new() }
+    }
+
+    #[cfg(not(test))]
+    fn empty() -> Self {
+        Self { extractors: Vec::new() }
+    }
+
+    /// Adds an extractor to the registry.
+    ///
+    /// Extractors are checked in the order they are added.
+    #[cfg(test)]
+    pub(crate) fn with_extractor<E: CallMetadataExtractor + 'static>(mut self, extractor: E) -> Self {
+        self.extractors.push(Arc::new(extractor));
+        self
+    }
+
+    #[cfg(not(test))]
+    fn with_extractor<E: CallMetadataExtractor + 'static>(mut self, extractor: E) -> Self {
+        self.extractors.push(Arc::new(extractor));
+        self
+    }
+
+    /// Analyzes the given context and returns a diagnostic if any extractor matches.
+    ///
+    /// Returns `None` if no extractor can handle the context.
+    pub async fn analyze(&self, context: &DiagnosticContext<'_>) -> Option<CallDiagnostic> {
+        for extractor in &self.extractors {
+            if extractor.can_handle(context) {
+                return Some(extractor.extract(context).await);
+            }
+        }
+        None
+    }
+
+    /// Analyzes and logs the diagnostic with structured fields.
+    ///
+    /// This is the primary method to use when handling transaction errors.
+    /// It extracts diagnostics and logs them with appropriate tracing spans
+    /// for CloudWatch/OTEL ingestion.
+    pub async fn analyze_and_log(&self, context: &DiagnosticContext<'_>) {
```

Comment:
Is there a case where we don't want to log ? The way the implementation is design it seems to me that it's not really a choice so why not keep it as "analyze", or "report" ?